npm/src/worker-proxy.ts

```diff
@@ -0,0 +1,387 @@
+      | { success: false; error: unknown };
+  };
+}[keyof T];
+
+/** Event message from the worker to the main thread */
+export type EventMessage<TEventMsg extends IServiceEventMessage> = {
+  messageType: "event";
+} & TEventMsg;
+
+/** Used as a constraint for events defined by the service */
+interface IServiceEventMessage {
+  type: string;
+  detail: unknown;
+}
+
+/**
+ * Strongly typed EventTarget interface. Used as a constraint for the
+ * event target that "requestWithProgress" methods should take in the service.
+ */
+export interface IServiceEventTarget<TEvents extends IServiceEventMessage> {
+  addEventListener<T extends TEvents["type"]>(
+    type: T,
+    listener: (event: Event & Extract<TEvents, { type: T }>) => void
+  ): void;
+
+  removeEventListener<T extends TEvents["type"]>(
+    type: T,
+    listener: (event: Event & Extract<TEvents, { type: T }>) => void
+  ): void;
+
+  dispatchEvent(event: Event & TEvents): boolean;
+}
+
+/** Holds state for a single request received by the proxy */
+type RequestState<
+  TService extends ServiceMethods<TService>,
+  TServiceEventMsg extends IServiceEventMessage
+> = RequestMessage<TService> & {
+  resolve: (val: any) => void;
+  reject: (err: any) => void;
+  requestEventTarget?: IServiceEventTarget<TServiceEventMsg>;
+  cancellationToken?: CancellationToken;
+};
+
+/*
+The WorkerProxy works by queuing up requests to send over to the Worker, only
+ever having one in flight at a time. By queuing on the caller side, this allows
+for cancellation (it checks if a request is cancelled before sending to the worker).
+
+The queue contains an entry for each request with the data to send, the promise
+to resolve, the event handler, and the cancellation token. When a request completes
+the next one (if present) is fetched from the queue. If it is marked as cancelled,
+it is resolved immediately, else it is marked as the current request and the command
+sent to the worker. As events occurs on the current request the event handler is
+invoked. When the response is received this is used to resolve the promise and
+complete the request.
+*/
+
+/**
+ * Function to create the proxy for a type. To be used from the main thread.
+ *
+ * @param postMessage A function to post messages to the worker
+ * @param terminator A function to call to tear down the worker thread
+ * @param methods A map of method names to be proxied and some metadata @see MethodMap
+ * @returns The proxy object. The caller should then set the onMsgFromWorker
+ * property to a callback that will receive messages from the worker.
+ */
+export function createProxy<
+  TService extends ServiceMethods<TService>,
+  TServiceEventMsg extends IServiceEventMessage
+>(
+  postMessage: (msg: RequestMessage<TService>) => void,
+  terminator: () => void,
+  methods: MethodMap<TService>
+): TService &
+  IServiceProxy & {
+    onMsgFromWorker: (
+      msg: ResponseMessage<TService> | EventMessage<TServiceEventMsg>
+    ) => void;
+  } {
```

Comment:
Wow. That's ummmm... quite the function signature :-o

Reply:
I promise it makes sense if you stare at it long enough :wink:

I think I can just create a type alias from the `onMsgFromWorker` function, that part looks messy.

- The type parameters `TService` and `TServiceEventMsg` , together with the `methods` parameters, describe the interface you want to proxy.
- `postMessage` and `terminator` are runtime-specific hooks.
- The return value just tells you it returns your original `TService` with a few additional methods unique to the proxy.

I don't love how complicated it looks, but I feel like every single parameter is _necessary_. But I may be overlooking something.

Reply:
I think it's fine. Don't spend more time on it. I'm sure we'll need to refactor all this worker stuff again at some point anyway.